test/compressor_test.go

```diff
@@ -39,6 +40,83 @@ import (
 	testpb "google.golang.org/grpc/interop/grpc_testing"
 )
 
+// TestUnsupportedEncodingResponse validates gRPC status codes for different client-server compression setups
+// ensuring the correct behavior when compression is enabled or disabled on either side.
+func (s) TestUnsupportedEncodingResponse(t *testing.T) {
+	tests := []struct {
+		name           string
+		clientUseNop   bool
+		serverUseNop   bool
+		expectedStatus codes.Code
+	}{
+		{
+			name:           "client_server_nop_compression",
+			clientUseNop:   true,
+			serverUseNop:   true,
+			expectedStatus: codes.OK,
+		},
+		{
+			name:           "client_nop_compression",
+			clientUseNop:   true,
+			serverUseNop:   false,
+			expectedStatus: codes.Unimplemented,
+		},
+		{
+			name:           "server_nop_compression",
+			clientUseNop:   false,
+			serverUseNop:   true,
+			expectedStatus: codes.Internal,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ss := &stubserver.StubServer{
+				UnaryCallF: func(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+					return &testpb.SimpleResponse{
+						Payload: in.Payload,
+					}, nil
+				},
+			}
+			sopts := []grpc.ServerOption{}
+			if test.serverUseNop {
+				// Using deprecated methods to selectively apply compression only on the server side.
+				// with encoding.registerCompressor(), the compressor is applied globally, affecting both the client and server.
+				sopts = append(sopts, grpc.RPCCompressor(newNopCompressor()), grpc.RPCDecompressor(newNopDecompressor()))
+			}
+			if err := ss.Start(sopts); err != nil {
+				t.Fatalf("Error starting server: %v", err)
+			}
+
+			defer ss.Stop()
+			dOpts := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
+			if test.clientUseNop {
+				// UseCompressor() requires the compressor to be registered using encoding.RegisterCompressor() which applies compressor globally,
+				// Hence, using deprecated WithCompressor() and WithDecompressor() to apply compression only on client.
+				dOpts = append(dOpts, grpc.WithCompressor(newNopCompressor()), grpc.WithDecompressor(newNopDecompressor()))
+			}
+			cc, err := grpc.NewClient(ss.Address, dOpts...)
+			if err != nil {
+				t.Fatalf("grpc.NewClient() returned unexpected error: %v", err)
+			}
+			defer cc.Close()
```

Comment:
You could replace these lines with a call to `ss.StartClient(dopts...)`, and then use `ss.Client.UnaryCall` to make the RPC.